### PR TITLE
[Builtins] Implement parsing of polymorphic data

### PR DESCRIPTION
**Pull request by effectfully**
Currently a huge mess, but seems to work.

### CODE REVIEW

File: plutus-core/plutus-core/src/PlutusCore/Parser.hs [hunk 1]
```diff
@@ -101,3 +101,18 @@ parseTerm = parseGen term
...
+-- >>> doIt term "(con bool True)"
+-- Right (Constant (SourcePos {sourceName = "test", sourceLine = Pos 1, sourceColumn = Pos 2}) (Some (ValueOf bool True)))
+-- >>> doIt term "(con (list bool) [True, False])"
+-- Right (Constant (SourcePos {sourceName = "test", sourceLine = Pos 1, sourceColumn = Pos 2}) (Some (ValueOf list (bool) [True,False])))
+-- >>> doIt term "(con (pair unit (list integer)) ((),[1,2,3]))"
+-- Right (Constant (SourcePos {sourceName = "test", sourceLine = Pos 1, sourceColumn = Pos 2}) (Some (ValueOf pair (unit) (list (integer)) ((),[1,2,3]))))
+-- >>> doIt term "(con (list (pair string bool)) [(\"abc\", True), (\"def\", False)])"
+-- Right (Constant (SourcePos {sourceName = "test", sourceLine = Pos 1, sourceColumn = Pos 2}) (Some (ValueOf list (pair (string) (bool)) [("abc",True),("def",False)])))
+-- >>> doIt term "(con string \"abc\")"
+-- Right (Constant (SourcePos {sourceName = "test", sourceLine = Pos 1, sourceColumn = Pos 2}) (Some (ValueOf string "abc")))
```
Comment (effectfully) [inline plutus-core/plutus-core/src/PlutusCore/Parser.hs:118]: Lists and pairs get parsed.

File: plutus-core/plutus-core/src/PlutusCore/Parser/Type.hs [hunk 2]
```diff
@@ -58,15 +67,45 @@ pType = choice $ map try
...
+-- >>> doIt defaultUniType "list integer"
+-- Right (SomeTypeIn (Kinded list (integer)))
+-- >>> doIt defaultUniType "pair (list bool)"
+-- Right (SomeTypeIn (Kinded pair (list (bool))))
+-- >>> doIt defaultUniType "pair (list unit) integer"
+-- Right (SomeTypeIn (Kinded pair (list (unit)) (integer)))
+-- >>> doIt defaultUniType "list (pair unit integer)"
+-- Right (SomeTypeIn (Kinded list (pair (unit) (integer))))
+-- >>> doIt defaultUniType "pair unit (pair bool integer)"
+-- Right (SomeTypeIn (Kinded pair (unit) (pair (bool) (integer))))
```
Comment (effectfully) [inline plutus-core/plutus-core/src/PlutusCore/Parser/Type.hs:111]: Built-in types of arbitrary kinds get parsed.

**Comment by kwxm** [comment]
Hmmm. I haven't looked at this yet, but I guess we'll have to add the syntax to the specification (it currently says we don't have any official syntax for these things). I'll add an issue.

**Comment by effectfully** [comment]
> Hmmm. I haven't looked at this yet, but I guess we'll have to add the syntax to the specification (it currently says we don't have any official syntax for these things). I'll add an issue.

Thank you. Don't worry about the syntax for now as it's not settled yet. What we have here should be good enough, though.

Note that this is not reviewable. It's basically a new starting point for the RV folks. It'll need polishing, tests and documentation before it makes sense for us to review it.

**Comment by kwxm** [comment]
> Note that this is not reviewable. It's basically a new starting point for the RV folks. It'll need polishing, tests and documentation before it makes sense for us to review it.

Yep.  I made an issue (PLT-378) before I forgot, but marked it as blocked until this was all finalised.

**Comment by effectfully** [comment]
Closing in favor of #4738.

**Pull request closed without merging**